### PR TITLE
SI-6969, mishandling of SoftReferences in method cache.

### PR DESCRIPTION
**Pull request by paulp**
More interesting to test than it was to fix. The soft
reference is now dereferenced once, the locally stored
underlying value ascertained to be non-null, and the
remainder of the references to the value use the local var.

The enclosed test reliably NPEs without this patch.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/CleanUp.scala [hunk 1]
```diff
@@ -228,16 +228,22 @@ abstract class CleanUp extends Transform with ast.TreeDSL {
...
+                    REF(reflPolyCacheSym) === gen.mkSoftRef(REF(methodCache))
+                  ) ENDIF,
+
+                  VAL(methodSym) === (REF(methodCache) DOT methodCache_find)(REF(forReceiverSym)),
+                  IF (REF(methodSym) OBJ_NE NULL) .
                     THEN (Return(REF(methodSym)))
                   ELSE {
                     def methodSymRHS  = ((REF(forReceiverSym) DOT Class_getMethod)(LIT(method), REF(reflParamsCacheSym)))
-                    def cacheRHS      = ((getPolyCache DOT methodCache_add)(REF(forReceiverSym), REF(methodSym)))
+                    def cacheRHS      = ((REF(methodCache) DOT methodCache_add)(REF(forReceiverSym), REF(methodSym)))
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/CleanUp.scala:252]: Could you please update the code in the [comment above](https://github.com/paulp/scala/blob/cec95d9f6ced49d79972a67fd029ccd964d51b38/src/compiler/scala/tools/nsc/transform/CleanUp.scala#L204)?

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/CleanUp.scala:252]: While you're at it, the code in the comment for the mono-cache is out of date with the implementation. That version doesn't appear (to me) to be susceptible to the same problem, but you might as well run the test for it as well.

Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/CleanUp.scala:252]: I'd rather tear it out. Do we have some evidence there is a population making an informed choice about the level of polymorphism in their structural type cache? I think that population should assume the maintenance requirements.

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/CleanUp.scala:252]: I concur. Just be sure to remove it from the command line opts (which stlll includes `invoke-dynamic`.)

**Comment by paulp** [comment]
Geez, how many single points of spurious failure is it reasonable to have?

PLS REBUILD ALL

**Comment by lrytz** [comment]
https://status.github.com/ is the issue this time

**Comment by paulp** [comment]
I know it's github, that's why I put it that way ("how many single points of failure")

**Comment by paulp** [comment]
PLS REBUILD ALL

**Comment by paulp** [comment]
(Rebased.)

**Comment by soc** [comment]
LGTM.
When this is fixed, I'll see whether this solves the issues for Avian and JamVM.

**Comment by paulp** [comment]
Great. I'll come back for mono type cache. Very interested to know if this advances the ball with avian.

**Pull request merged**